src/java/org/apache/cassandra/index/sai/plan/VectorTopKProcessor.java

```diff
@@ -101,44 +141,58 @@ public <U extends Unfiltered, R extends BaseRowIterator<U>, P extends BasePartit
         // to store top-k results in primary key order
         TreeMap<PartitionInfo, TreeSet<Unfiltered>> unfilteredByPartition = new TreeMap<>(Comparator.comparing(p -> p.key));
 
-        while (partitions.hasNext())
-        {
-            try (R partition = partitions.next())
-            {
-                DecoratedKey key = partition.partitionKey();
-                Row staticRow = partition.staticRow();
-                PartitionInfo partitionInfo = PartitionInfo.create(partition);
-                // compute key and static row score once per partition
-                float keyAndStaticScore = getScoreForRow(key, staticRow);
+        if (PARALLEL_STAGE != Stage.IMMEDIATE && partitions instanceof ParallelizablePartitionIterator) {
+            ParallelizablePartitionIterator pIter = (ParallelizablePartitionIterator) partitions;
+            var commands = pIter.getUninitializedCommands();
+            List<CompletableFuture<List<Triple<PartitionInfo, Row, Float>>>> results = new LinkedList<>();
 
-                while (partition.hasNext())
-                {
-                    Unfiltered unfiltered = partition.next();
-                    // Always include tombstones for coordinator. It relies on ReadCommand#withMetricsRecording to throw
-                    // TombstoneOverwhelmingException to prevent OOM.
-                    if (!unfiltered.isRow())
+            int count = commands.size();
+            for (var command: commands) {
+                CompletableFuture<List<Triple<PartitionInfo, Row, Float>>> future = new CompletableFuture<>();
+                results.add(future);
+
+                // run last command immediately, others in parallel (if possible)
+                count--;
+                var executor = count == 0 ? Stage.IMMEDIATE : PARALLEL_STAGE;
+
+                executor.maybeExecuteImmediately(() -> {
```

Comment:
is it meaningful to call this instead of execute()?  the method is not documented but it looks to me like it is only relevant to the single-threaded executor

Reply:
It is documented in LocalAwareExecutorService. 
LAES takes a parameter (called num threads/max concurrency/max pool size interchangeably in different classes) which works as limit on a number of tasks that can be submitted there, while the actual threads are shared with other executors in the pool. When the limit is reached, maybeExecuteImmediately will run the task on the current thread.